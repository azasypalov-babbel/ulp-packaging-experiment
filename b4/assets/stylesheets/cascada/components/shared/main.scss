@use "sass:math";
// Layout

%responsive-container {
  margin: 0 auto;
  width: 90%;

  @include media(medium) {
    max-width: 940px;
  }
  @include media(large) {
    // fixed width for large and beyond
    max-width: none;
    width: 940px;
  }
  @include media(extra-large) {
    width: 1170px;
  }
}

// Typography

%trainer-page-text {
  font-size: math.div(16, 15) * 1em;
  margin-bottom: rem(24px);
  text-align: center;
}

%centered-columns {
  box-sizing: border-box;
  display: flex;
  justify-content: center;
  flex-direction: column;
}

// UI

%bordered-box {
  border-radius: $border-radius;
  background-color: $white;
  box-shadow: 0 2px 2px 0 rgba(0, 0, 0, 0.16),
              0 0 2px 0 rgba(0, 0, 0, 0.12),
              inset 1px 1px 0 0 rgba(0, 0, 0, 0.08);
  color: $space-gray;
}

// Cascada override for Safari in iOS 10

.btn--primary,
.btn {
  display: inline-block;
}
