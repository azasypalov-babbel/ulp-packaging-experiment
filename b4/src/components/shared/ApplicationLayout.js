import styled, { css } from 'styled-components';
import { isiOS, isWebview } from '../../lib/features';

const fixNavigationBar = css`
  & .loy .loy-cascada-navbar {
    position: fixed;
    top: 0;
    padding-top: env(safe-area-inset-top);

    &:after {
      top: initial;
    }
  }
  & > main {
    margin-top: calc(${(props) => props.theme.size.navbar.xsmall.height} + env(safe-area-inset-top));
  }
`;

const disableSelection = css`
  user-select: none;
  -webkit-touch-callout: none;
`;

const ApplicationLayout = styled.div`
  display: flex;
  flex-direction: column;

  @media all and (-ms-high-contrast: none), (-ms-high-contrast: active) {
    height: 100vh; /* IE11 does not respect min-height for calculating flex */
  }

  min-height: 100vh;
  ${isiOS() && css`
    min-height: -webkit-fill-available; /* To reduce the page size if possible when software keyboard is open */
  `}

  ${isWebview() && fixNavigationBar};
  ${isWebview() && disableSelection};

  & > main {
    position: relative;
    flex: 1 0 auto; /* main should be least the height of it's content */

    display: flex;

    & > * {
      flex: 1 1 100%; /* make children 100% width, and shrink if screen width is too small */
    }
  }

  .hide-for-tiny {
    @media (max-width: 413px) {
      display: none;
    }
  }

  .fixed-for-small {
    @media (max-width: 767px) {
      position: fixed;
    }
  }

  .hide-for-small {
    @media (max-width: 767px) {
      display: none;
    }
  }
`;

export default ApplicationLayout;
