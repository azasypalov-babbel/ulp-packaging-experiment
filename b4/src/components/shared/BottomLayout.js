import React, { useEffect, useLayoutEffect, useState } from 'react';
import { getDisplayName } from '../getDisplayName';
import PropTypes from 'prop-types';
import styled from 'styled-components';
import { createPortal } from 'react-dom';
import useResizeObserver from './hooks/useResizeObserver';
import { getBottomInset, scrollIntoSafeArea, setBottomInset } from './scrollIntoSafeArea';
import { connect } from 'react-redux';

/**
 * This area is for any elements that are fixed to the bottom of the page
 * Give components `order` css property to organise them.
 */

const mountPointId = 'bottom-layout-mount-point';

export const StyledBottomLayout = styled.div`
  position: relative;

  @supports (position: sticky) {
    position: sticky;
    height: ${(props) => props.internalHeight}px;
  }

  bottom: 0;
  left: 0;
  right: 0;

  flex: 0 0 auto;
  z-index: 99;

  pointer-events: none;

  & > * {
    @supports (position: sticky) {
      position: absolute;
    }

    left: 0;
    right: 0;
    bottom: 0;
    display: flex;
    flex-direction: column;

    pointer-events: all;
  }
`;

export const bottomLayoutConfig = {
  enableScrollAdjustment: true,
  enableShrinking: false
};

const BottomLayoutComponent = ({ children, id }) => {
  const [internalHeight, setInternalHeight] = useState(0);

  const onResize = ({ height }) => {
    const isGrowing = height > internalHeight;

    if (isGrowing) {
      /**
       * InternalHeight is used to reserve space in the DOM.
       * It can only increase in height, to avoid making the page jump.
       */
      setInternalHeight(height);
    } else if (height !== 0 && bottomLayoutConfig.enableShrinking) {
      setInternalHeight(height);
    }

    const shouldAdjustScroll = height > getBottomInset();

    /**
     * Height of this element is used to calculate a safe area for scrolling
     * ie. so that content is not covered by anything in BottomLayout
     */
    setBottomInset(height);

    if (shouldAdjustScroll && bottomLayoutConfig.enableScrollAdjustment) {
      /**
       * Delay scroll until after next paint so that elements are in their final place
       */
      requestAnimationFrame(() => scrollIntoSafeArea());
    }
  };

  useEffect(() => {
    /**
     * Reset internal height when a the trainer changes or sequence starts or ends
     */
    setInternalHeight(0);
  }, [id]);

  const elementRef = useResizeObserver(onResize);

  return (
    <StyledBottomLayout internalHeight={internalHeight}>
      <div
        ref={elementRef}
        id={mountPointId}
      >
        {children}
      </div>
    </StyledBottomLayout>
  );
};

BottomLayoutComponent.propTypes = {
  children: PropTypes.node,
  id: PropTypes.string
};

const mapStateToProps = ({ sequence }) => ({
  id: [sequence.currentTrainerIndex, sequence.started, sequence.completed].join()
});

export const BottomLayout = connect(mapStateToProps)(BottomLayoutComponent);

export const renderInBottomLayout = (WrappedComponent) => {
  const RenderInBottomLayout = (props) => {
    const [element, setElement] = useState(document.getElementById(mountPointId));

    useLayoutEffect(() => {
      const domNode = document.getElementById(mountPointId);
      if (!domNode) {
        // eslint-disable-next-line max-len
        console.warn(`DOM element with id ${mountPointId} not found. This is a requirement for a bottom layout to be properly rendered.`);
      }
      setElement(domNode);
    }, []);

    if (element === null) {
      return null;
    }

    return createPortal(
      <WrappedComponent {...props} />,
      element
    );
  };

  RenderInBottomLayout.displayName = `RenderInBottomLayout(${getDisplayName(WrappedComponent)})`;

  return RenderInBottomLayout;
};
