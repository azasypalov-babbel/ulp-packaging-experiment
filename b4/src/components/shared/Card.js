import { transparentize, rem } from 'polished';
import styled, { css } from 'styled-components';

const Card = styled.div`
  box-sizing: border-box;
  overflow: hidden;

  background: ${(props) => props.theme.cascada.white};

  border-radius: ${rem('4px')};

  transition: box-shadow 0.3s ease;

  box-shadow: 0 ${rem('2px')} ${rem('20px')} 0 ${(props) => transparentize(0.9, props.theme.cascada.spaceGrayW15)};

  outline: none;

  ${(props) => Boolean(props.onClick) && css`

    cursor: pointer;

    &:hover, &:focus {
      box-shadow: 0 ${rem('8px')} ${rem('40px')} 0 ${(props) => transparentize(0.8, props.theme.cascada.spaceGrayW15)};
    }

    &:active {
      box-shadow: 0 ${rem('2px')} ${rem('10px')} 0 ${(props) => transparentize(0.7, props.theme.cascada.spaceGrayW15)};
    }
  `}
`;

export default Card;
