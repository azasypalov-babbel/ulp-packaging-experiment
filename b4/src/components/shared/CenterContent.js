import styled from 'styled-components';

const CenterContent = styled.div`
   display: flex;
   align-items: center;

   & > * {
      margin: auto;
   }
`;

export default CenterContent;
