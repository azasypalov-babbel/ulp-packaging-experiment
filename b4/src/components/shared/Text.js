import PropTypes from 'prop-types';
import styled from 'styled-components';

import * as features from '../../lib/features';

const fontFamily = ({ fontFamily, theme }) =>
  theme.cascada[fontFamily] || theme.cascada.fontMilliardBook;

const color = ({ color, theme }) =>
  theme.cascada[color] || 'inherit';

const textAlign = ({ textAlign }) =>
  textAlign || 'inherit';

const getFontSize = (theme, sizeName) =>
  theme.size.font[sizeName] || 'inherit';

const Text = styled.span`
  font-size: ${({ fontSize = {}, theme }) => {
    const fontSizeName = typeof fontSize === 'string'
      ? fontSize
      : fontSize.mobile;
    return getFontSize(theme, fontSizeName);
  }};
  font-family: ${fontFamily};
  color: ${color};
  font-weight: normal;
  line-height: 1.25;
  letter-spacing: 0.031rem;
  text-align: ${textAlign};
  ${() => {
    if (features.isWebview()) {
      return 'user-select: none;';
    }
  }};
  ${({ fontSize, theme }) => {
    if (fontSize && fontSize.desktop) {
      return `
        @media (min-width: 768px) {
          font-size: ${getFontSize(theme, fontSize.desktop)};
        }`;
    }
  }}

  & > strong,
  & > b {
    font-family: ${(props) => props.theme.cascada.fontMilliardSemi};
    font-weight: normal;
  }

  & > em,
  & > i {
    font-family: ${(props) => props.theme.cascada.fontMilliardItalic};
    font-weight: normal;
    font-style: normal;
  }

  & > strong em,
  & > b i {
    font-family: ${(props) => props.theme.cascada.fontMilliardMediumItalic};
    font-weight: normal;
  }
`;

const fontSizeNames = [
  'xsmall',
  'small',
  'base',
  'medium',
  'big',
  'large',
  'xlarge',
  'grand'
];

Text.propTypes = {
  fontSize: PropTypes.oneOfType([
    PropTypes.oneOf(fontSizeNames),
    PropTypes.shape({
      desktop: PropTypes.oneOf(fontSizeNames),
      mobile: PropTypes.oneOf(fontSizeNames).isRequired
    })
  ]),
  fontFamily: PropTypes.oneOf([
    'fontMilliardSemi',
    'fontMilliardBook',
    'fontMilliardMedium',
    'fontLeituraBold'
  ]),
  textAlign: PropTypes.oneOf(['left', 'center', 'right']),
  color: PropTypes.string,
  verticalAlign: PropTypes.string
};

export default Text;
