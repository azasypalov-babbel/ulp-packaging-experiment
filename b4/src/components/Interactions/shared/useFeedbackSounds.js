import { useCallback, useContext, useEffect, useReducer, useRef } from 'react';
import { ServiceContext } from '@src/components/shared/withServices';
import incorrectFeedbackSound from '@assets/sounds/babbel_wrong.mp3';
import correctFeedbackSound from '@assets/sounds/babbel_correct.mp3';
import { useSoundPlayer } from '@src/components/Trainer/shared/useSoundPlayer';
import { checks } from '@src/services/soundService';

const useFeedbackSounds = ({ skipOnActiveSound = true, resetDependencies = [] } = {}) => {
  const { settingsService, soundService } = useContext(ServiceContext);
  const correctSoundInstance = useSoundPlayer(correctFeedbackSound);
  const incorrectSoundInstance = useSoundPlayer(incorrectFeedbackSound);
  const [, forceUpdate] = useReducer((s) => !s, true);
  const skippedFeedbackRef = useRef(false);

  useEffect(() => {
    skippedFeedbackRef.current = false;
    correctSoundInstance.reset();
    incorrectSoundInstance.reset();
  }, [correctSoundInstance, incorrectSoundInstance, ...resetDependencies]);

  const hasPlayed = checks.hasPlayed(correctSoundInstance.getState())
    || checks.hasPlayed(incorrectSoundInstance.getState());

  return [
    useCallback((solved) => {
      const isPlayingAnySound = soundService.getPlayingInstances().length > 0;
      if ((skipOnActiveSound && isPlayingAnySound) || !settingsService.isFeedbackSoundEnabled) {
        skippedFeedbackRef.current = true;
        forceUpdate();
        return;
      }
      const instance = solved ? correctSoundInstance : incorrectSoundInstance;
      instance.play();
    }, [skipOnActiveSound, soundService, settingsService, correctSoundInstance, incorrectSoundInstance]),
    hasPlayed || skippedFeedbackRef.current
  ];
};

export default useFeedbackSounds;
