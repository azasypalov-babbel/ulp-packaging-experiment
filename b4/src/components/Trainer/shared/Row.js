import React, { useEffect, useRef } from 'react';
import PropTypes from 'prop-types';
import styled, { keyframes, css } from 'styled-components';
import { rem } from 'polished';
import { scrollIntoSafeArea } from '../../shared/scrollIntoSafeArea';

const enterKeyframes = keyframes`
  to {
    opacity: 0.05;
  }
`;

const RowBase = styled.div`
  position: relative;
  animation: ${enterKeyframes} 0.3s ease-out ${({ delayIndex }) => (delayIndex * 0.1) + 0.5}s reverse backwards;

  padding-bottom: ${rem('6px')};
  scroll-padding: ${rem('12px')};

  & + & {
    margin-top: ${rem('12px')};
    padding-top: ${rem('6px')};
  }
`;

RowBase.propTypes = {
  delayIndex: PropTypes.number
};

const StyledTask = styled(RowBase)`
  & + & {
    margin-top: 0;
  }
`;

const StyledPhrase = styled(RowBase)`
  padding-left: ${rem('12px')};
  border-left: ${rem('4px')} solid ${(props) => props.theme.color.edge.raw.yellow.joyful};

  & + & {
    margin-top: 0;
    padding-top: ${rem('12px')};
  }

  ${({ onClick }) => onClick !== null ? css`&:hover { cursor: pointer; }` : null};
`;

const StyledExample = styled.div`
  display: flex;
  flex-direction: column;

  & > * + * {
    margin-top: ${rem('4px')};
  }
`;

const PhraseRow = React.forwardRef((props, ref) => (
  <StyledPhrase ref={ref} {...props}>
    <StyledExample>
      {props.children}
    </StyledExample>
  </StyledPhrase>
));

PhraseRow.displayName = 'PhraseRow';

PhraseRow.propTypes = {
  children: PropTypes.node.isRequired,
  onClick: PropTypes.func
};

export default PhraseRow;


export const Row = ({ type, active, delayIndex = 0, ...props }) =>{
  const ref = useRef();
  useEffect(() => {
    if (active) {
      scrollIntoSafeArea(ref.current);
    }
  }, [ref, active]);

  return (type === 'task'
    ? <StyledTask
      ref={ref}
      delayIndex={delayIndex}
      data-item-type="task"
      {...props}
    />
    : <PhraseRow
      ref={ref}
      delayIndex={delayIndex}
      data-item-type="phrase"
      {...props}
    />);
};

Row.displayName = 'Row';

Row.propTypes = {
  active: PropTypes.bool,
  type: PropTypes.oneOf(['task', 'phrase']),
  delayIndex: PropTypes.number
};
