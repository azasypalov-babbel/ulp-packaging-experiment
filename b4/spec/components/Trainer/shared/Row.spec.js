import React from 'react';
import Text from '../../../../src/components/shared/Text';
import { shallow } from 'enzyme';
import { Row } from '../../../../src/components/Trainer/shared/Row';

const children = <Text>A cold day in L.A</Text>;

const props = {
  children,
  translation: 'Un dia frio en Los Angeles'
};

describe('<Row />', () => {
  describe('task row', () => {
    test('it renders', () => {
      const wrapper = shallow(<Row type="task" {...props} />);
      expect(wrapper).toMatchSnapshot();
    });
  });

  describe('phrase row', () => {
    test('it renders', () => {
      const wrapper = shallow(<Row type="phrase" {...props} />);
      expect(wrapper).toMatchSnapshot();
    });
  });
});
