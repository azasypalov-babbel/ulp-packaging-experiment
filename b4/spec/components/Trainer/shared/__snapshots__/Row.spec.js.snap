// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Row /> phrase row it renders 1`] = `
<PhraseRow
  data-item-type="phrase"
  delayIndex={0}
  translation="Un dia frio en Los Angeles"
>
  <styled.span>
    A cold day in L.A
  </styled.span>
</PhraseRow>
`;

exports[`<Row /> task row it renders 1`] = `
<Styled(styled.div)
  data-item-type="task"
  delayIndex={0}
  translation="Un dia frio en Los Angeles"
>
  <styled.span>
    A cold day in L.A
  </styled.span>
</Styled(styled.div)>
`;
